### ISSUE TITLE
sys-usb sometimes fails to start

### ISSUE THREAD

**Issue by hackerfactor** [QubesOS/qubes-issues#6824]
### Qubes OS version
R4.1beta
/etc/qubes-release: Qubes release 4.1 (R4.1)

### Affected component(s) or functionality
sys-usb, mouse

-----
### Brief summary
This appears to be a race condition. It does not happen all the time.

1. Install Qubes 4.1 (Qubes-R4.1.0-beta1-x86_64.iso).
2. During the install, I deselect Whonix (don’t need it for my use model) and I toggle sys-firewall and sys-usb to NOT be disposible. (For my needs, I customize sys-usb.)
3. Reboot. sys-usb does not consistently start up.
4. Open Qube Manager. Select settings for sys-usb, toggle start on boot (was checked, uncheck it, then recheck it). Click “Apply” and “OK”. From this point on, sys-usb appears to start properly during boot.

The results that I see:
Most of the time (about 80%), sys-usb starts properly.
About 15% of the time, sys-usb appears to be running, but the mouse (touch pad on laptop) does not respond. Using keystrokes, I can get to the shutdown menu and reboot the system.
About 5% of the time, even the keyboard doesn't work. In this case, plugging in USB devices does not even trigger any notification messages. It appears that sys-usb failed to start -- or it started, but failed to acquire the usb bus.

One time, I had no keyboard or mouse. After about 30 seconds of sitting idle, I saw a bunch of notifications as every USB bus and device was recognized. But the keyboard and mouse were still dead.

The laptop is a Dell 5550.

Other people have reported the same problem:
https://forum.qubes-os.org/t/sys-usb-not-starting-at-boot/5336/3
https://forum.qubes-os.org/t/sys-usb-not-starting-at-boot/5336/7

To me, this looks like a race condition. sys-usb is not starting up with the hardware is ready, or is not accepting hardware if it appears during the startup.

### How reproducible
For me, it's happening about 20% of the time.

### Steps to reproduce
See bug description.

### Expected behavior
sys-usb should start up every time. Keyboard and mouse should always work.

### Actual behavior
See bug description.

### Solutions you've tried
Tried disposible and non-disposible USB.

R4.0 on the same laptop has no problems. (Never saw this issue.) It's only in R4.1.

### Additional context
The most frustrating part is that it is not easily reproducable. It really acts like a startup race condition.

-----
### Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted
Discussions on the forums: https://forum.qubes-os.org/t/sys-usb-not-starting-at-boot/5336

### Related, [non-duplicate](https://www.qubes-os.org/doc/issue-tracking/#new-issues-should-not-be-duplicates-of-existing-issues) issues
The forum has a report about PCI devices not starting up at boot:
https://forum.qubes-os.org/t/after-logging-in-complete-black-screen-in-4-1/5210
If it's a race condition with PCI devices, then it could impact mouse, keyboard, and screen.

**Comment by hackerfactor** [comment]
I think I have a way to recreate this.  (Worked 3 times in a row.)

During the boot, you prompted to enter the LUKS password.
Before entering any password -- just let it sit at the LUKS password prompt for about a minute. (E.g., boot the laptop but then get distracted by a phone call or cute cat video on a different computer.)  Then enter the password to unlock the device.  The laptop will boot but the mouse will be dead.

If I do it quickly: boot / password, then the mouse works.

**Comment by Ark-of-the-copepod** [comment]
FWIW, my solution is to use a keybinding to start up sys-usb when that happens.

Long-standing annoyance.

**Comment by hackerfactor** [comment]
> FWIW, my solution is to use a keybinding to start up sys-usb when that happens.

Can you provide more information about where and how to do this?
Specifically: Is this added to dom0 or sys-usb? Where is it added (what file or startup script)? And what is the actual command-line?

**Comment by unman** [comment]
As another data point, I regularly see this error with 4.1
What is interesting is that I can manually start sys-usb once the system
is running.

Looking at the logs, I see two entries of interest:
1. Configuration file ***@***.***/50_autostart.conf is marked world-inaccessible.
2. A libvirtd error - failed to create new domain 'sys-usb', unable to reset PCI device 0000:00:14.0

The first isn't important.
The second is. I'm not clear on *why* that device needs to be reset, and
why that fails on some occasions. Nor why it can be started later
without issue.

**Comment by marmarek** [comment]
This is indeed very interesting. Can you check `/var/log/libvirt/libxl/libxl-driver.log` for more details about this? And maybe `/var/log/xen/console/guest-sys-usb-dm.log` (or whatever usbvm name you have), but likely there won't be much info there (I'm hoping for qemu error message).

**Comment by Ark-of-the-copepod** [comment]
@hackerfactor : No problem.

Find the 'Keyboard' option in the Q menu (start menu in dom0), choose 'Application Shortcuts'. In there you can set key-bindings to launch applications but it just writes a dom0 commandline input when you press the shortcut so I use it for this.

```qvm-start sys-usb```
```qvm-shutdown sys-usb```

I also use this to get Qubes to display my screens properly. 50-75% of the time it forgets on boot it has 2 screens, so a shortcut is easier than
```xrandr --output VGA1 --primary --auto --output LVDS1 --auto --pos 200x1050```

Caution- There may be a better way of doing things, my knowledge is low.

**Comment by unman** [comment]
On Tue, Aug 24, 2021 at 09:00:37AM -0700, Marek Marczykowski-Górecki wrote:
> This is indeed very interesting. Can you check `/var/log/libvirt/libxl/libxl-driver.log` for more details about this? And maybe `/var/log/xen/console/guest-sys-usb-dm.log` (or whatever usbvm name you have), but likely there won't be much info there (I'm hoping for qemu error message).
> 

in `/var/log/libvirt/libxl/libxl-driver.log` i see:
```
libxl: libxl_pci.c:1493:libxl_device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0
libxl: libxl_pci.c:1772:libxl_device_pci_add_done: Domain 3:libxl_device_pci_add failed for device 0000:00:14.0
```

I can post full log if needed.

**Comment by hackerfactor** [comment]
Found it!

If Qubes boots and the mouse does not run, you can restart the mouse driver on dom0:

1. Find the mouse driver:  lsmod | grep touch
Mine says "hid_multitouch  32768 0"

2. Remove it:  sudo rmmod hid_multitouch
3. Add it back: sudo modprobe hid_multitouch

Since this problem happens way too often for me (1 out of every 4 reboots), I've added this to $HOME/.config/xfce4/xinitrc
  mouse=$(lsmod | grep touch | awk '{print $1}')
  sudo rmmod "$mouse"
  sudo modprobe "$mouse"
This way, I bounce it every time the system boots -- whether it needs it or not.

(I'm still doing lots of reboots, but this seems to be working.)

**Comment by marmarek** [comment]
Interesting, is this driver actually connected to your mouse? I don't think it is a _mouse_ driver...
You can check it this way:
```
ls -l /sys/module/hid_multitouch/drivers/*/
```
While I don't exclude possibility this driver somehow interfere with your mouse, it should not be necessary for normal operation (even if you have a touchscreen). Can you confirm by not loading it back? If so, maybe simply blacklist it via `/etc/modprobe.conf.d`?

**Comment by hackerfactor** [comment]
@marmarek
Without my workaround in xinitrc, the mouse is dead about 1 in 4 reboots. (Mouse cursor is on the screen, but touchpad does nothing.)

If I "sudo rmmod hid_multitouch", then the mouse stays dead.
If I then use "sudo modprobe hid_multitouch", then the mouse immediately comes alive.

It's a trackpad built into the laptop.  (Dell 5550)

**Comment by marmarek** [comment]
For a built-in device, it is unlikely to have anything to do with sys-usb then...

**Comment by hackerfactor** [comment]
I'm suspecting that it has more to do with the PCI being ready.  Maybe it starts slow or doesn't always respond to the first reset? 

That would cause sys-usb to sometimes die -- race condition when the PCI is unavailable when sys-usb starts up.
It would also cause multitouch to fail -- race condition when the PCI isn't ready during modprobe but comes up before sys-usb is started.

**Comment by hackerfactor** [comment]
Update: I think I have a fix.  I'm doing this in dom0:$HOME/.config/xfce4/xinitrc

```
# Bounce mouse to ensure it always works. Qubes race condition, bug https://github.com/QubesOS/qubes-issues/issues/6824
lsmod | grep touch | while read a b ; do
  sudo rmmod "$a"
  sudo modprobe "$a"
done

# Check if sys-usb is running. Qubes race condition, bug https://github.com/QubesOS/qubes-issues/issues/6824
SysUSB=$(qvm-ls --raw-data --fields STATE sys-usb)
if [ "$SysUSB" == "Halted" ] ; then
  qvm-start sys-usb
fi
```
So far, no more dead mice after multiple reboots.

**Comment by 3hhh** [comment]
> 2021-08-26 15:38:42.039+0000: libxl: libxl_pci.c:1493:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0

Got the same issue here in 4.1rc2, even for exactly the same device (Intel USB Hub). It happens even though no-strict-reset=True is set for it. Sometimes the same happens for sys-net Intel Wifi.

I guess no-strict-reset=True is broken in 4.1rc2.

It works fine in 4.0 and IIRC even in 4.1rc1.

**Comment by Szewcson** [comment]
I also noticed it. It is relay annoying. Sometimes after boot only thing I can do is shutdown since my keyboard (internal USB) and mouse (external USB dongle) not works. Also it appears not only after boot. It also happens sometimes when woken up or returning from suspend...

**Comment by Szewcson** [comment]
What is the current status of this issue?

**Comment by DemiMarie** [comment]
I am not aware of any changes.

**Comment by ddevz** [comment]
If this is a separate issue, let me know and i'll create a new issue out of it, but since its sys-usb race condition and might be related, i'll start with it here:

Sometimes sys-usb *does* start, but shows no usb devices in sys-usb when using "lsusb"
When this happens, it seems to be *all 5* sys-usb qubes that have the problem (once sys-usb qube per usb controller).
They also may be failing to start at all some of the time.  (will document this better the next time it happens)


Possibly related:
Someone mentioned a timing related correlation with how long you sit at the LUKS password prompt.  
I've created this part as a separate issue, https://github.com/QubesOS/qubes-issues/issues/7434 , but basically:
I see a LUKS password prompt, but typing shows no feedback acknowledging that you have entered anything (for example showing a dot for each character entered).  
After typing in the LUKS password and hitting enter, the screen does not change.   Sometimes after a while (30+ seconds) it will disappear and the user login screen will present

**Comment by turmoil995** [comment]
Same for me, from time to time bultin touchpad doesn't work in q 4.1 kernel 5.10.104-3.fc32.qubes.x86_64 built-in keyboard works and I'm able at least make a restart. Also in /var/log/libvirt/libxl/libxl-driver.log I get:
`
2022-05-01 20:00:31.551+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:0d.0
2022-05-01 20:00:31.555+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:0d.2
2022-05-01 20:00:31.560+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0
2022-05-01 20:00:42.692+0000: libxl: libxl_pci.c:1774:device_pci_add_done: Domain 3:libxl__device_pci_add  failed for PCI device 0:0:d.2 (rc -9)
2022-05-01 20:00:42.695+0000: libxl: libxl_pci.c:1774:device_pci_add_done: Domain 3:libxl__device_pci_add  failed for PCI device 0:0:14.0 (rc -9)
2022-05-01 20:00:42.695+0000: libxl: libxl_create.c:1904:domcreate_attach_devices: Domain 3:unable to add pci devices
2022-05-01 20:00:42.704+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:0d.0
2022-05-01 20:00:42.855+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:0d.2
2022-05-01 20:00:42.989+0000: libxl: libxl_pci.c:2019:do_pci_remove: Domain 4:xc_physdev_unmap_pirq irq=16: Invalid argument
2022-05-01 20:00:42.990+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0
2022-05-01 20:00:43.256+0000: libxl: libxl_exec.c:120:libxl_report_child_exitstatus: killed fork (dying as expected) [6079] unexpectedly exited status zero
The kernel doesn't support reset from sysfs for PCI device 0000:00:0d.2
`
About hit_multitouch:
removing it during the normal operation (when everything starts properly - most of the time) results in the touchpad strange behaviour (unexpected action such as clicks/drags etc). But I'm able to move cursor.

**Comment by DemiMarie** [comment]
That is not surprising.  Use no_strict_reset for the USB controller.

**Comment by marmarek** [comment]
And if you have already no-strict-reset set (see `qvm-pci` command) - which is done by default for sys-usb, check /var/log/xen/console/guest-sys-usb-dm.log - there may be more details about this failed PCI attach.

**Comment by mfc** [comment]
i faced the same issues after upgrading Heads firmware. error in ` /var/log/libvirt/libxl/libxl-driver.log` is same:
`libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0`

running `sudo qubesctl state.sls qvm.sys-usb` did not fix this, which is surprising if this is supposed to be done by default for sys-usb.

i needed to set `no-strict-reset` on *all* PCI devices attached to `sys-usb`, not just the PCI device that has that error.

**Comment by DemiMarie** [comment]
> i needed to set `no-strict-reset` on _all_ PCI devices attached to `sys-usb`, not just the PCI device that has that error.

If you set `no-strict-reset` on the one with that error, do you then get an error about a different device?

**Comment by mfc** [comment]
> > i needed to set `no-strict-reset` on _all_ PCI devices attached to `sys-usb`, not just the PCI device that has that error.
> 
> If you set `no-strict-reset` on the one with that error, do you then get an error about a different device?

no same device. i used GUI to select (`sys-usb` Qube Settings > Devices > Configure strict reset for PCI devices), maybe using command-line works as expected.

**Comment by DemiMarie** [comment]
@mfc can you make a separate bug with the details?

**Comment by ddevz** [comment]
Confirmation on relationship between bugs plus proposed interm "solution" (at end):

Even though I know if I reboot and type the LUKS password in quicker, I grabbed some data to try to help with the issue

**relationship between bugs**
It looks to me that both 'sys-usb sometimes fails to start' and 'Sometimes sys-usb does start, but shows no usb devices in sys-usb when using "lsusb"' are both caused by a race condition that happens depending on how long one sits at the LUKS prompt

Normally I experience the 'Sometimes sys-usb does start, but shows no usb devices in sys-usb when using "lsusb"' issue when I let it sit at the LUKS prompt for more then somewhere in the 60 second range or so.  However this time, I got the 'sys-usb sometimes fails to start' problem.  The big difference is that a power failure happened, causing me to come back to a computer that was already at the LUKS prompt.

From dmesg, we can see that the LUKS prompt sat there for approximately 267339 seconds (I.E. 3 days)

[   11.720123] [drm] Initialized radeon 2.50.0 20080528 for 0000:b3:00.0 on minor 0
[267350.098945] kauditd_printk_skb: 8 callbacks suppressed
[267350.098946] audit: type=1130 audit(1654008572.804:25): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-cryptsetup@luks{{some stuff i'm removing it case it's the password in the log}} comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[267350.953886] audit: type=1130 audit(1654008573.659:26): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=dracut-initqueue comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[267351.148996] audit: type=1130 audit(1654008573.854:27): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-fsck-root comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[267351.168273] EXT4-fs (dm-4): mounted filesystem with ordered data mode. Opts: (null)

Hopefully knowing that these are related can help in tracking this down

**Proposed interm solution:**

I propose that until this gets tracked down, that a note be added to the LUKS password screen that mentions that waiting at that screen can cause some people USB issues.  Possibly even waiting and only displaying that after 20 seconds if they haven't gotten through the LUKS prompt already.

(I don't know how difficult making UI changes that early in the boot process would be)

**Comment by DemiMarie** [comment]
That is _strange_.  I have no idea why sitting for a long time at the LUKS prompt could cause USB issues.  Does it hit some sort of timeout in the hardware, where the hardware expects the OS to initialize it within a certain amount of time, and fails when this does not happen?

**Comment by DemiMarie** [comment]
If so, this will be fixed by bringing up sys-usb from the initramfs, but I do not see any other way to solve the problem.

**Comment by ddevz** [comment]
No idea if this is related to anything, but in my dmesg from this time (when things are working), I have:
systemd[1]: Listening on LVM2 poll daemon socket

Yet, the dmesg from the time it did not work there is no such line.

(found using `sed 's/[^]]*]//'` and `diff --side-by-side`)

**Comment by 3hhh** [comment]
Btw I hvane't had this issue anymore for a while and also the random reboots on sys-usb startup disappeared for now.

For me it might have been related to a Xen or dom0 kernel update.

I still experience the aforementioned mousepad/cursor issue, but I guess that's unrelated as it's not in sys-usb. Also it usually fixes itself after a few seconds.

**Comment by tlikonen** [comment]
Just adding some info: This happens to me most of the time. Sometimes sys-usb starts automatically but more often I have to start it manually. My computer is Nitropad T430 (i.e., certified hardware) and Qubes 4.1. In the installation I chose to configure sys-usb as disposable.

**Comment by ddevz** [comment]
> Just adding some info: This happens to me most of the time. Sometimes sys-usb starts automatically but more often I have to start it manually. My computer is Nitropad T430 (i.e., certified hardware) and Qubes 4.1. In the installation I chose to configure sys-usb as disposable.

How many VM's do you have set to autostart on boot?

**Comment by tlikonen** [comment]
> How many VM's do you have set to autostart on boot?
Just the default basic ones: sys-net, sys-firewall and sys-usb.

**Comment by mfc** [comment]
i just encountered this again on a reboot, even though i have `no-strict-reset` configured on all devices attached to `sys-usb`.

`/var/log/libvirt/libxl/libxl-driver.log`:
```
2022-06-19 09:58:22.166+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0
2022-06-19 09:58:34.108+0000: libxl: libxl_pci.c:1774:device_pci_add_done: Domain 3:libxl__device_pci_add  failed for PCI device 0:0:1d.0 (rc -9)
2022-06-19 09:58:34.134+0000: libxl: libxl_pci.c:1774:device_pci_add_done: Domain 3:libxl__device_pci_add  failed for PCI device 0:0:14.0 (rc -9)
2022-06-19 09:58:34.134+0000: libxl: libxl_create.c:1904:domcreate_attach_devices: Domain 3:unable to add pci devices
2022-06-19 09:58:35.183+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0
2022-06-19 09:59:56.026+0000: libxl: libxl_pci.c:1489:libxl__device_pci_reset: The kernel doesn't support reset from sysfs for PCI device 0000:00:14.0
```

after manually running `qvm-start sys-usb`:

`qvm-pci list`:
```
BACKEND:DEVID  DESCRIPTION                                                                                                        USED BY
dom0:00_00.0   Host bridge: Intel Corporation 3rd Gen Core processor DRAM Controller                                              
dom0:00_02.0   VGA compatible controller: Intel Corporation 3rd Gen Core processor Graphics Controller                            
dom0:00_04.0   Signal processing controller: Intel Corporation 3rd Gen Core Processor Thermal Subsystem                           
dom0:00_14.0   USB controller: Intel Corporation 7 Series/C210 Series Chipset Family USB xHCI Host Controller                     sys-usb (no-strict-reset=True)
dom0:00_16.0   Communication controller: Intel Corporation 7 Series/C216 Chipset Family MEI Controller #1                         
dom0:00_19.0   Ethernet controller: Intel Corporation 82579LM Gigabit Network Connection (Lewisville)                             sys-net
dom0:00_1a.0   USB controller: Intel Corporation 7 Series/C216 Chipset Family USB Enhanced Host Controller #2                     sys-usb (no-strict-reset=True)
dom0:00_1b.0   Audio device: Intel Corporation 7 Series/C216 Chipset Family High Definition Audio Controller                      
dom0:00_1c.0   PCI bridge: Intel Corporation 7 Series/C216 Chipset Family PCI Express Root Port 1                                 
dom0:00_1c.1   PCI bridge: Intel Corporation 7 Series/C210 Series Chipset Family PCI Express Root Port 2                          
dom0:00_1c.2   PCI bridge: Intel Corporation 7 Series/C210 Series Chipset Family PCI Express Root Port 3                          
dom0:00_1d.0   USB controller: Intel Corporation 7 Series/C216 Chipset Family USB Enhanced Host Controller #1                     sys-usb (no-strict-reset=True)
dom0:00_1f.0   ISA bridge: Intel Corporation QM77 Express Chipset LPC Controller                                                  
dom0:00_1f.2   SATA controller: Intel Corporation 7 Series Chipset Family 6-port SATA Controller [AHCI mode]                      
dom0:00_1f.3   SMBus: Intel Corporation 7 Series/C216 Chipset Family SMBus Controller                                             
dom0:00_1f.6   Signal processing controller: Intel Corporation 7 Series/C210 Series Chipset Family Thermal Management Controller  
dom0:01_00.0   SD Host controller: Ricoh Co Ltd PCIe SDXC/MMC Host Controller                                                     sys-usb (no-strict-reset=True)
dom0:02_00.0   Network controller: Qualcomm Atheros AR93xx Wireless Network Adapter                                               sys-net
```

**Comment by ddevz** [comment]
> That is _strange_. I have no idea why sitting for a long time at the LUKS prompt could cause USB issues. Does it hit some sort of timeout in the hardware, where the hardware expects the OS to initialize it within a certain amount of time, and fails when this does not happen?

New information on this:
When installing a new system, apparently the new version of the installer does not allow you to create a sys-usb qube when it finds fake USB keyboards and fake USB storage devices. :) (this is now common on high end systems as part of "remote management").  (I get it, no matter what you do for the sys-usb defaults, someone is going to run into trouble)
Anyway so i did the "create a sys-usb" salt script thing, and after I did that, the next boot did not display the "stars" when entering the LUKS password.
I suspect this is because of the "pretend that the USB devices do not exist for dom0" boot parameter that gets added by the salt script  that makes the sys-usb qube.  (it's probably this: 
    GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX rd.qubes.hide_all_usb"
)

Is it possible that the "pretend that the USB devices do not exist for dom0" boot parameter is too hiding something that some video cards use?  If so this could also potentially explain the video artifacts that some people see.

Hope this helps

**Comment by tlaurion** [comment]
> **Automated announcement from [builder-github](https://github.com/QubesOS/qubes-builder-github)**
> 
> The component `vmm-xen` (including package `python3-xen-4.14.5-8.fc32`) has been pushed to the `r4.1` testing repository for dom0. To test this update, please install it with the following command:
> 
> ```
> sudo qubes-dom0-update --enablerepo=qubes-dom0-current-testing
> ```
> 
> [Changes included in this update](https://github.com/QubesOS/qubes-vmm-xen/compare/24ca3023d57ece2a412548a9c7cb028f852610dc...1b7f208dd7d42435bad111a7234dedeaaff5f43c)

`sudo qubes-dom0-update --disablerepo=* --enablerepo=qubes-dom0-current-testing`

Results upon reboot into sys-usb not failing to start randomly anymore (x230). After sys-usb restart, issue gone as well.
